### PR TITLE
NIFI-5249 Dockerfile enhancements

### PR DESCRIPTION
**Pull request by pepov**
- Createdata repository and log directories
- Add explicit environment variable declarations to replace nifi-env.sh
- Remove obsolete chown commands after user has already been switched to nifi

How I tested the changes:
- I've ran `mvn package -P docker` in the nifi-docker folder to build the image. 
- I've made sure that `docker run --rm -ti apache/nifi` runs and start nifi using the new image
- Made sure that all files are owned by the nifi user under /opt/nifi: `docker run --rm -ti --entrypoint /bin/bash apache/nifi -c "find /opt/nifi ! -user nifi"`
- Checked that the expected environment variables are set:
```
docker run --rm -ti --entrypoint /bin/bash apache/nifi -c "env | grep NIFI"
NIFI_HOME=/opt/nifi/nifi-1.7.0-SNAPSHOT
NIFI_LOG_DIR=/opt/nifi/nifi-1.7.0-SNAPSHOT/logs
NIFI_PID_DIR=/opt/nifi/nifi-1.7.0-SNAPSHOT/run
NIFI_BASE_DIR=/opt/nifi
```

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by ottobackwards** [comment]
Can you add explicit steps to test and verify, for those of use who don't work with docker as much?

**Comment by pepov** [comment]
yes, working on it

**Comment by ottobackwards** [comment]
Thanks, I'd like to try this and help with the review

**Comment by pepov** [comment]
Now all the above described test steps can be executed with `mvn package -P docker` from the `nifi-docker` folder

**Comment by ottobackwards** [comment]
So to test this?
- mvn package
- mvn package -P docker from nifi-docker
- ???
- docker run --rm -ti --entrypoint /bin/bash apache/nifi -c "env | grep NIFI" ? from nifi-docker dir?
- docker run --rm -ti --entrypoint /bin/bash apache/nifi -c "find /opt/nifi ! -user nifi"  from nifi-docker dir?

**Comment by pepov** [comment]
No, sorry! I meant integration-test not package. I beleive this would do it, at least works for me, just tested:
```
mvn package -pl nifi-assembly
cd nifi-docker
mvn integration-test -P docker
```

**Comment by ottobackwards** [comment]
+1 fwiw|lgtm.
Ran the above steps to build and test the image from the maven snapshots ( not built locally ).
Everything ran fine.

Your integration test is awesome.  I'm totally going to steal it.

Super stuff.

File: nifi-docker/dockermaven/pom.xml [hunk 1]
```diff
@@ -44,7 +44,6 @@
                                     </buildArgs>
                                     <repository>apache/nifi</repository>
                                     <tag>${project.version}</tag>
-                                    <tag>latest</tag>
```
Comment (zenfenan) [inline nifi-docker/dockermaven/pom.xml]: Why was this removed? When we do a release, I believe we want to keep that release as the latest.

Comment (pepov) [reply nifi-docker/dockermaven/pom.xml]: I removed it because it did not build both tags, just the last one, so I had a choice. Also this is only a dev/test config and does not affect the public image.

File: nifi-docker/dockerhub/Dockerfile [hunk 2]
```diff
@@ -25,28 +25,37 @@ ARG GID=1000
...
     && apt-get update \
-    && apt-get install -y jq xmlstarlet
+    && apt-get install -y jq xmlstarlet procps
 
 USER nifi
 
 # Download, validate, and expand Apache NiFi binary.
 RUN curl -fSL ${MIRROR}/${NIFI_BINARY_URL} -o ${NIFI_BASE_DIR}/nifi-${NIFI_VERSION}-bin.tar.gz \
-    && echo "$(curl https://archive.apache.org/dist/${NIFI_BINARY_URL}.sha256) *${NIFI_BASE_DIR}/nifi-${NIFI_VERSION}-bin.tar.gz" | sha256sum -c - \
+    && echo "$(curl ${MIRROR}/${NIFI_BINARY_URL}.sha256) *${NIFI_BASE_DIR}/nifi-${NIFI_VERSION}-bin.tar.gz" | sha256sum -c - \
```
[the remark this answers is not in the document]
Comment (apiri) [reply nifi-docker/dockerhub/Dockerfile]: @pepov It is an automated build in Docker Hub.  The Dockerfile is used as shown without any external arguments when a rel/nifi-<version> tag is created

[the remark this answers is not in the document]
Comment (pepov) [reply nifi-docker/dockerhub/Dockerfile]: I've restored the hardcoded url

**Comment by pepov** [comment]
Rebased and squashed commits

**Comment by pepov** [comment]
Are there any issues left I did not address?

**Comment by pepov** [comment]
To reply to my own question: I think we should also test the dockerhub image almost the same way we test the maven one. We just have to use the latest public nifi archive to do that.

File: nifi-docker/dockermaven/pom.xml (hunk 1 above)
Comment (zenfenan) [reply nifi-docker/dockermaven/pom.xml]: Understood. Thanks.

File: nifi-docker/dockermaven/Dockerfile [hunk 3]
```diff
@@ -26,23 +26,33 @@ ARG NIFI_BINARY
...
+# Web HTTP(s) & Socket Site-to-Site Ports
+EXPOSE 8080 8443 10000
 
-WORKDIR $NIFI_HOME
+WORKDIR ${NIFI_HOME}
 
 # Startup NiFi
 ENTRYPOINT ["bin/nifi.sh"]
-CMD ["run"]
+CMD ["run"]
```
Comment (jtstorck) [inline nifi-docker/dockermaven/Dockerfile]: Creating a container with:
`docker run -p 8080:8080 apache/nifi:1.7.0-SNAPSHOT-dockermaven`
results in NiFi starting successfully, but I'm unable to control-c out of the container.  I'm not a docker expert, but I would expect that hitting control-c would kill the container.  Although, since I didn't run it in interactive mode, this is probably a docker thing, not specific to this Dockerfile.

Starting the container with:
`docker run -d -p 8080:8080 apache/nifi:1.7.0-SNAPSHOT-dockermaven`
and then issuing:
`docker logs <container>`
I see the nifi-bootstrap output, but not the nifi-app.log and nifi-user.log output.  Would it be preferable to have this behavior?

Comment (pepov) [reply nifi-docker/dockermaven/Dockerfile]: Ctrl-C works for me, even if it's somewhat slow while nifi is booting (faster when already booted).
The dockermaven image does not contain the wrapper scripts that the dockerhub one does, this is why tailing and sending all logs to stdout doesn't work. I wasn't sure that was intentional or not but I can add it to make the two environments more similar.

Comment (jtstorck) [reply nifi-docker/dockermaven/Dockerfile]: I tried to use Ctrl-C after NiFi was successfully up and running to kill the container.  I had to open a new shell and use docker kill to bring it down.  Not a big deal, I still think that's due to docker and having started the container in a non-interactive and non-detached method.

For the logging, I'm not sure if there's a reason why the two docker modules have different wrapper scripts.  You could check with @apiri, but most likely it'd be good to bring them in line.  It doesn't have to be done in this PR, though it'd be nice to get this into the NiFi 1.7.0 release if no one disagrees with it.

Comment (pepov) [reply nifi-docker/dockermaven/Dockerfile]: You're right I totally missed that. The biggest problem is that it is even the case with the dockerhub image, which is much more painful. I think I know what the problem is and working on the fix.
